scripts/dataset_processing/tts/compute_speaker_stats.py

```diff
@@ -0,0 +1,108 @@
+
+$ python <nemo_root_path>/scripts/dataset_processing/tts/compute_speaker_stats.py \
+    --manifest_path=<data_root_path>/fastpitch_manifest.json \
+    --sup_data_path=<data_root_path>/sup_data \
+    --pitch_stats_path=<data_root_path>/pitch_stats.json
+"""
+
+import argparse
+import json
+import os
+from collections import defaultdict
+from pathlib import Path
+from typing import List, Tuple
+
+import torch
+from tqdm import tqdm
+
+from nemo.collections.tts.data.data_utils import get_sup_data_file_path, read_manifest
+from nemo.collections.tts.torch.helpers import get_base_dir
+from nemo.collections.tts.torch.tts_data_types import Pitch
+from nemo.utils import logging
+
+
+def get_args():
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter, description="Compute speaker level pitch statistics .",
+    )
+    parser.add_argument(
+        "--manifest_path", required=True, type=Path, help="Path to training manifest.",
+    )
+    parser.add_argument(
+        "--sup_data_path", default=Path("sup_data"), type=Path, help="Path to base directory with supplementary data.",
+    )
+    parser.add_argument(
+        "--pitch_stats_path",
+        default=Path("pitch_stats.json"),
+        type=Path,
+        help="Path to output JSON file with speaker pitch statistics.",
+    )
+    args = parser.parse_args()
+    return args
+
+
+def _compute_stats(values: List[torch.Tensor]) -> Tuple[float, float]:
+    values_tensor = torch.cat(values, dim=0)
+    mean = values_tensor.mean().item()
+    std = values_tensor.std(dim=0).item()
+    return mean, std
+
+
+def main():
+    args = get_args()
+    manifest_path = args.manifest_path
+    sup_data_path = args.sup_data_path
+    pitch_stats_path = args.pitch_stats_path
+
+    entries = read_manifest(manifest_path)
+
+    audio_paths = [entry["audio_filepath"] for entry in entries]
+    base_dir = get_base_dir(audio_paths)
+
+    global_pitch_values = []
+    speaker_pitch_values = defaultdict(list)
+    for entry in tqdm(entries):
+        pitch_path = get_sup_data_file_path(entry, base_dir, sup_data_path, Pitch.name)
+        if not os.path.exists(pitch_path):
+            logging.warning(f"Unable to find pitch file for {entry}")
+            continue
+
+        pitch = torch.load(pitch_path)
+        # Filter out non-speech frames
+        pitch = pitch[pitch != 0]
+        global_pitch_values.append(pitch)
+        if "speaker" in entry:
+            speaker_id = entry["speaker"]
+            speaker_pitch_values[speaker_id].append(pitch)
+
+    global_pitch_mean, global_pitch_std = _compute_stats(global_pitch_values)
+    pitch_stats = {"default": {"pitch_mean": global_pitch_mean, "pitch_std": global_pitch_std}}
+    for speaker_id, pitch_values in speaker_pitch_values.items():
+        pitch_mean, pitch_std = _compute_stats(pitch_values)
+        pitch_stats[speaker_id] = {"pitch_mean": pitch_mean, "pitch_std": pitch_std}
```

Comment:
why not use the same stats to calculate `pitch_min` and `pitch_max` ?

Reply:
Do you mean include pitch_min and pitch_max just for informational purposes? Or to replace the "pitch_fmin" and "pitch_fmax" values in the config? It cannot do the latter because those are not used at train time, but as inputs to librosa.pyin() to compute the F0 to begin with.

Reply:
Yes I mean pitch_fmin and pitch_fmax. They are used in the `__get_item__` method here https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/torch/data.py#L538
I don't follow why can't we get these also from your dict?

Reply:
The fmin and fmax are not dataset statistics, they are bounds (hardcoded to 65Hz-2093Hz) to make the F0 estimation more computationally efficient.

That line which calls "librosa.pyin()" is only called if you compute the pitch on the first epoch instead of running "extract_sup_data.py". This script computes the pitch mean/std using the outputs of librosa.pyin(), assuming it was done beforehand using "extract_sup_data.py".

Though that also means no one should use this branch of the data loader you linked, as that would mean they begin training before knowing the pitch norm statistics. Some of our past tutorials suggested people do this without realizing the bug (causing them to train on their dataset using the hardcoded LJSpeech mean/std).

Reply:
I see now our FP training tutorial suggests tuning fmin/fmax to target speaker, which is very confusing. The only reason to tune it (if I am not mistaken) is if you want to (1) reduce the range to make the compute faster, but risk truncating the pitch values or (2) have training data greater than 2k Hz such as singing or yelling, and are willing to heavily slow down the compute to capture those ranges.

https://github.com/NVIDIA/NeMo/blame/main/tutorials/tts/FastPitch_Finetuning.ipynb#L347-L349

Reply:
Reaching back in my memory--I recall we had some issues where some speakers we were fine-tuning on had noisier environments, and pyin kept predicting pitches for static/noise as speech. So in the end we went speaker-by-speaker and adjusted fmin (don't think we touched fmax) to avoid "noise-as-speech" with minimal loss of actual speech pitches.

Anyway, that's probably what that line refers to, but yeah... without more explanation it's somewhat unclear.

Reply:
That would make sense. You might be able to eliminate outliers or influence the VAD output by restricting the range of F0 to normal human speech (I have no clue how the VAD in pyin works).